[BLS] Add wrapper around chiabls lib, worker, benchmarks and unit tests

random-zebra opened the pull request:
Based on top of:
- [x] #2419 

This reworks #2405, adapting the commits to the new build system (which includes the library already at `v1.0.1`, as git subtree).

random-zebra commented:
Rebased on master. Ready for review.

furszy reviewed (commented):
Code reviewed up to 32121d35.
So far, so good ☕ ☕. Left two comments in the BLS IES area.

Plus added test coverage for the basic encryption and decryption scheme here https://github.com/furszy/PIVX/commit/8b8c223394aa34dee65f4e21443f071e5dfab2b2. If you want to cherry-pick It, all yours.

First stop before jumping into the thread pool workers and DKG bench waters.

random-zebra commented on src/bls/bls_ies.cpp:
```diff
@@ -0,0 +1,131 @@
+    for (size_t i = 0; i < count; i++) {
+        ivVector[i] = iv;
+        iv = ::SerializeHash(iv);
+    }
```
:+1: Updated.

random-zebra commented on src/bls/bls_ies.cpp:
```diff
@@ -0,0 +1,131 @@
+#include "streams.h"
+
+template <typename Out>
+static bool EncryptBlob(const void* in, size_t inSize, Out& out, const void* symKey, const void* iv)
+{
+    out.resize(inSize);
+
+    AES256CBCEncrypt enc((const unsigned char*)symKey, (const unsigned char*)iv, false);
+    int w = enc.Encrypt((const unsigned char*)in, (int)inSize, (unsigned char*)out.data());
+    return w == (int)inSize;
```
I wouldn't add an exception for this, as it is not "exceptional behavior". That is what the return value is for (and the caller must always check it).
If padding is disabled and the data size is not a multiple of the aes blocksize, then `AES256CBCEncrypt::Encrypt` returns 0, and this function returns false for any (non empty) object.

Also, keep in mind that this function, as well as the `CBLSIESEncryptedBlob` class, is low level. We don't access them directly, but with templated wrappers (`CBLSIESEncryptedObject<T>` / `CBLSIESMultiRecipientObjects<T>`), where the size is fixed, and given by the serialized size of the underlying object `T` (for starters a BLS secret key, which has size 32).

`CBLSIESMultiRecipientObjects` is the class we are going to use in the DKG, to exchange encrypted key shares between quorum members.
`CBLSIESEncryptedObject` is not used anywhere, actually, and I think it was added only for completeness. 
We could even consider dropping it from the codebase, and then adding it back if/when there's a specific use-case for it.

furszy replied:
k yeah.
it's more for an internal base class documentation to not have to re-discover this in the future in case we decide to use it for a different purpose. The not padding decision is done internally without any comment nor description.
 But, we can just add a descriptive comment at the top of the base class in the header and done for now as well.

random-zebra replied:
Added few comments.

random-zebra commented:
Thanks @furszy . Rebased on master and picked your test, only reworking it using `CBLSIESEncryptedObject<std::string>` instead of `CBLSIESEncryptedBlob` directly (in case we want to keep this code).

Inspired by this, have also added encryption/decryption (with `CBLSIESMultiRecipientObjects<CBLSSecretKey>`) for the contribution shares in the DKG test.

furszy reviewed (commented):
Reviewed till a24d9bad01372b3295dd16634a680a6684ac95b5.
Few more comments, nothing biggie.

random-zebra commented on src/bls/bls_worker.cpp:
```diff
@@ -0,0 +1,962 @@
+
+void CBLSWorker::Start()
+{
+    int workerCount = std::thread::hardware_concurrency() / 2;
```
I've replaced it here. But wouldn't make `mt_pooled_secure.h` dependent on `util/system.h` just for this...

furszy commented on src/test/bls_tests.cpp:
```diff
@@ -51,4 +52,124 @@ BOOST_AUTO_TEST_CASE(bls_sig_tests)
+            rcvContributions.emplace_back(quorum[j].contributions[i]);
+        }
+        m.skShare = worker.AggregateSecretKeys(rcvContributions);
+        allSkShares.emplace_back(m.skShare);
```
This vector `allSkShares` is not being used, can remove it.

random-zebra replied:
yup :+1: leftover from a previous implementation

furszy commented on src/test/bls_tests.cpp:
```diff
@@ -51,4 +52,124 @@ BOOST_AUTO_TEST_CASE(bls_sig_tests)
+        for (size_t j = 0; j < N; j++) {
+            rcvContributions.emplace_back(quorum[j].contributions[i]);
+        }
+        m.skShare = worker.AggregateSecretKeys(rcvContributions);
```
Can verify the validity of this sk share below with a:

```
// Verify share validity
BLSPublicKeyVector aggrVerifVector;
for (size_t j = 0; j < N; j++) {
    aggrVerifVector.emplace_back(worker.BuildPubKeyShare(quorum[j].vecP, m.id));
}
CBLSPublicKey aggrPkShare = worker.AggregatePublicKeys(aggrVerifVector);
BOOST_CHECK(m.skShare.GetPublicKey() == aggrPkShare);
```

random-zebra replied:
Mhm... Yeah. 
This is redundant, as it is implied by the above `VerifyContributionShare` (the pk, corresponding to the share sent by member `j` to member `m`, is the evaluation point of `j`'s P-vector at `m`'s id, i.e. the output of `BuildPubKeyShare`), and by the aggregation properties of BLS keys (the public key of the secret key, resulting from the aggregation of N secret keys, is equal to the aggregation of the corresponding N public keys).

But still... we can add this test, so we exercise CBLSWorker's `BuildPubKeyShare` and `AggregatePublicKeys` functions too.

furszy reviewed (commented):
Review completed, the DKG test made me feel at home ☕ . Code ACK 21d2f247d760bf748f83c5d154addf0b3c44edd4.
Left few more comments, nothing blocking. Can squash the changes and it's ready to go for me.

random-zebra commented:
Rebased and squashed.

furszy reviewed (approved):
one step closer, ACK 0d126eee5e1cb88ff8ae3cd2d7ee6a8920f82978

Fuzzbawls reviewed (approved):
ACK 0d126eee5e1cb88ff8ae3cd2d7ee6a8920f82978
